### PR TITLE
Add a an ast.Walk utility function

### PR DESCRIPTION
**Pull request by rs**
Add an `ast.Walk` utility function to traverse the AST in depth-first order.

### CODE REVIEW

File: ast/walk.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+// Visitor documentation.
+type VisitorFunc func(Visitor, Node) Visitor
+
+// Visit method is invoked for each node encountered by Walk.
+func (vf VisitorFunc) Visit(n Node) Visitor {
+	return vf(vf, n)
+}
+
+// Walk traverses an AST in depth-first order: It starts by calling
+// v.Visit(node); node must not be nil. If the visitor w returned by
```
Comment (stevenh) [inline ast/walk.go]: Confusion between v and w in the docs and code

File: ast/walk.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
+package ast
+
+import "fmt"
+
+// Visitor Visit method is invoked for each node encountered by Walk.
+// If the result visitor w is not nil, Walk visits each of the children
+// of node with the visitor w, followed by a call of w.Visit(nil).
+type Visitor interface {
+	Visit(n Node) (w Visitor)
```
Comment (stevenh) [inline ast/walk.go]: v would be more appropriate

**Comment by rs** [comment]
Fixed.

File: ast/walk.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+			Walk(v, n.Test)
+			Walk(v, n.Body)
+		}
+	case *WithStatement:
+		if n != nil {
+			Walk(v, n.Object)
+			Walk(v, n.Body)
+		}
+	default:
+		panic(fmt.Sprintf("Walk: unexpected node type %T", n))
```
Comment (stevenh) [inline ast/walk.go]: Would this be better as an error instead of a panic?

Comment (rs) [inline ast/walk.go]: If this condition happen, it means we missed on type so it is a programing error. The Go's walker does the same: https://github.com/golang/go/blob/master/src/go/ast/walk.go#L364

**Pull request merged**